// @udecode/plate-code-block/dist/server.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/server.ts
var server_exports = {};
__export(server_exports, {
  CODE_BLOCK_LANGUAGES: () => CODE_BLOCK_LANGUAGES,
  CODE_BLOCK_LANGUAGES_POPULAR: () => CODE_BLOCK_LANGUAGES_POPULAR,
  ELEMENT_CODE_BLOCK: () => ELEMENT_CODE_BLOCK,
  ELEMENT_CODE_LINE: () => ELEMENT_CODE_LINE,
  ELEMENT_CODE_SYNTAX: () => ELEMENT_CODE_SYNTAX
});
module.exports = __toCommonJS(server_exports);

// src/shared/constants.ts
var ELEMENT_CODE_BLOCK = "code_block";
var ELEMENT_CODE_LINE = "code_line";
var ELEMENT_CODE_SYNTAX = "code_syntax";
var CODE_BLOCK_LANGUAGES_POPULAR = {
  bash: "Bash",
  css: "CSS",
  git: "Git",
  graphql: "GraphQL",
  html: "HTML",
  javascript: "JavaScript",
  json: "JSON",
  jsx: "JSX",
  markdown: "Markdown",
  sql: "SQL",
  svg: "SVG",
  tsx: "TSX",
  typescript: "TypeScript",
  wasm: "WebAssembly"
};
var CODE_BLOCK_LANGUAGES = {
  antlr4: "ANTLR4",
  bash: "Bash",
  c: "C",
  cmake: "CMake",
  coffeescript: "CoffeeScript",
  csharp: "C#",
  css: "CSS",
  dart: "Dart",
  django: "Django",
  docker: "Docker",
  ejs: "EJS",
  erlang: "Erlang",
  git: "Git",
  go: "Go",
  graphql: "GraphQL",
  groovy: "Groovy",
  html: "HTML",
  java: "Java",
  javascript: "JavaScript",
  json: "JSON",
  jsx: "JSX",
  kotlin: "Kotlin",
  latex: "LaTeX",
  less: "Less",
  lua: "Lua",
  makefile: "Makefile",
  markdown: "Markdown",
  markup: "Markup",
  matlab: "MATLAB",
  objectivec: "Objective-C",
  perl: "Perl",
  php: "PHP",
  powershell: "PowerShell",
  properties: ".properties",
  protobuf: "Protocol Buffers",
  python: "Python",
  r: "R",
  ruby: "Ruby",
  sass: "Sass (Sass)",
  // scala: 'Scala',
  scheme: "Scheme",
  // FIXME: Error with current scala grammar
  scss: "Sass (Scss)",
  shell: "Shell",
  sql: "SQL",
  svg: "SVG",
  swift: "Swift",
  tsx: "TSX",
  typescript: "TypeScript",
  wasm: "WebAssembly",
  xml: "XML",
  yaml: "YAML"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CODE_BLOCK_LANGUAGES,
  CODE_BLOCK_LANGUAGES_POPULAR,
  ELEMENT_CODE_BLOCK,
  ELEMENT_CODE_LINE,
  ELEMENT_CODE_SYNTAX
});
